src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -371,8 +362,9 @@ static void GetHashCode<TValue>(ref HashCode hc, ConfigurationList<TValue> list)
             }
 
             // An options instance might be locked but not initialized for reflection serialization yet.
+            // Ensure hashing is consistent before and after rooting by returning null when the default resolver is used.
             private static IJsonTypeInfoResolver? NormalizeResolver(IJsonTypeInfoResolver? resolver)
-                => resolver ?? DefaultJsonTypeInfoResolver.DefaultInstance;
+                => resolver == DefaultJsonTypeInfoResolver.DefaultInstance ? null : resolver;
```

Comment:
why do we need to change this? either collating to null or default seems fine, or are you worried about multi-threaded? if so the same problem will exist either way

Reply:
The current behavior is actually a bug, since it will result in inconsistent hashes being generated depending on whether the default resolver has been rooted or not.